File: sdks/go/pricing.go
package costobs

import (
	_ "embed"
	"strings"
	"sync"

	"github.com/shopspring/decimal"
	"gopkg.in/yaml.v3"
)

//go:embed data/pricing-v2026.06.yaml
var pricingYAML []byte

type pricingFile struct {
	Version string `yaml:"version"`
	Models  []struct {
		Provider           string   `yaml:"provider"`
		Model              string   `yaml:"model"`
		Match              string   `yaml:"match"`
		Operation          string   `yaml:"operation"`
		InputPerToken      *float64 `yaml:"input_per_token"`
		CachedInputPerTok  *float64 `yaml:"cached_input_per_token"`
		OutputPerToken     *float64 `yaml:"output_per_token"`
		ReasoningPerToken  *float64 `yaml:"reasoning_per_token"`
		ToolPerToken       *float64 `yaml:"tool_per_token"`
		AudioPerSecond     *float64 `yaml:"audio_per_second"`
		PerCharacter       *float64 `yaml:"per_character"`
		BatchDiscount      *float64 `yaml:"batch_discount"`
		FinetuneSurcharge  *float64 `yaml:"finetune_surcharge"`
		ImageTiers         []struct {
			PerTile float64 `yaml:"per_tile"`
			Base    float64 `yaml:"base"`
		} `yaml:"image_tiers"`
	} `yaml:"models"`
}

type modelRule struct {
	provider, model, match string
	input, cachedInput, output, reasoning, tool *decimal.Decimal
	audioPerSecond, perCharacter                *decimal.Decimal
	imagePerTile, imageBase                     *decimal.Decimal
}

func (r *modelRule) matches(provider, model string) bool {
	if r.provider != provider {
		return false
	}
	if r.match == "prefix" {
		return strings.HasPrefix(model, r.model)
	}
	return model == r.model
}

// PricingEngine computes per-call cost from the bundled pricing file using
// exact decimal math. Unknown models cost 0 until pricing is added.
type PricingEngine struct {
	Version string
	rules   []modelRule

	mu    sync.RWMutex
	cache map[string]*modelRule
}

var (
	defaultEngine     *PricingEngine
	defaultEngineOnce sync.Once
)

// DefaultPricing returns the shared engine loaded from the embedded file.
func DefaultPricing() *PricingEngine {
	defaultEngineOnce.Do(func() {
		eng, err := NewPricingEngine(pricingYAML)
		if err != nil {
			// The embedded file is validated by tests; an error here means a
			// broken build, so fall back to an engine that prices everything 0.
			eng = &PricingEngine{Version: "invalid", cache: map[string]*modelRule{}}
		}
		defaultEngine = eng
	})
	return defaultEngine
}

// NewPricingEngine parses a pricing YAML document.
func NewPricingEngine(raw []byte) (*PricingEngine, error) {
	var pf pricingFile
	if err := yaml.Unmarshal(raw, &pf); err != nil {
		return nil, err
	}
	dec := func(p *float64) *decimal.Decimal {
		if p == nil {
			return nil
		}
		d := decimal.NewFromFloat(*p)
		return &d
	}
	eng := &PricingEngine{Version: pf.Version, cache: map[string]*modelRule{}}
	for _, m := range pf.Models {
		r := modelRule{
			provider:       m.Provider,
			model:          m.Model,
			match:          m.Match,
			input:          dec(m.InputPerToken),
			cachedInput:    dec(m.CachedInputPerTok),
			output:         dec(m.OutputPerToken),
			reasoning:      dec(m.ReasoningPerToken),
			tool:           dec(m.ToolPerToken),
			audioPerSecond: dec(m.AudioPerSecond),
			perCharacter:   dec(m.PerCharacter),
		}
		if len(m.ImageTiers) > 0 {
			t := decimal.NewFromFloat(m.ImageTiers[0].PerTile)
			b := decimal.NewFromFloat(m.ImageTiers[0].Base)
			r.imagePerTile, r.imageBase = &t, &b
		}
		eng.rules = append(eng.rules, r)
	}
	return eng, nil
}

func (e *PricingEngine) find(provider, model string) *modelRule {
	key := provider + "\x00" + model
	e.mu.RLock()
	if r, ok := e.cache[key]; ok {
		e.mu.RUnlock()
		return r
	}
	e.mu.RUnlock()

	var found *modelRule
	// Exact matches take priority; then longest prefix wins.
	for i := range e.rules {
		r := &e.rules[i]
		if r.match != "prefix" && r.matches(provider, model) {
			found = r
			break
		}
	}
	if found == nil {
		for i := range e.rules {
			r := &e.rules[i]
			if r.match == "prefix" && r.matches(provider, model) {
				if found == nil || len(r.model) > len(found.model) {
					found = r
				}
			}
		}
	}
	e.mu.Lock()
	e.cache[key] = found
	e.mu.Unlock()
	return found
}

// Cost computes the USD cost of one call. Returns 0 for unknown models.
func (e *PricingEngine) Cost(u Usage, provider, model string) decimal.Decimal {
	rule := e.find(provider, model)
	if rule == nil {
		return decimal.Zero
	}
	total := decimal.Zero
	mulInt := func(rate *decimal.Decimal, n int) {
		if rate != nil && n != 0 {
			total = total.Add(rate.Mul(decimal.NewFromInt(int64(n))))
		}
	}
	mulInt(rule.input, u.InputTokens)
	if u.CachedInputTokens > 0 {
		rate := rule.cachedInput
		if rate == nil {
			rate = rule.input
		}
		mulInt(rate, u.CachedInputTokens)
	}
	mulInt(rule.output, u.OutputTokens)
	if u.ReasoningTokens > 0 {
		rate := rule.reasoning
		if rate == nil {
			rate = rule.output
		}
		mulInt(rate, u.ReasoningTokens)
	}
	mulInt(rule.tool, u.ToolTokens)
	if u.AudioSeconds > 0 && rule.audioPerSecond != nil {
		total = total.Add(rule.audioPerSecond.Mul(decimal.NewFromFloat(u.AudioSeconds)))
	}
	mulInt(rule.perCharacter, u.Characters)
	if u.ImageTiles > 0 && rule.imagePerTile != nil {
		mulInt(rule.imagePerTile, u.ImageTiles)
		if u.ImageCount > 0 && rule.imageBase != nil {
			mulInt(rule.imageBase, u.ImageCount)
		}
	}
	return total
}

File: sdks/python/costobs/pricing.py
"""Synchronous local cost calculation from the bundled pricing file.

The pricing YAML is loaded once at import time and held in memory. All
arithmetic uses :class:`decimal.Decimal` for exactness — float rounding is
unacceptable for money.
"""

from __future__ import annotations

import threading
from decimal import Decimal
from typing import Dict, List, Optional, Tuple

from .schema import Usage

_PRICING_PKG = "costobs.data"
_PRICING_FILE = "pricing-v2026.06.yaml"


def _load_pricing_text() -> str:
    """Read the bundled pricing YAML via importlib.resources (wheel-safe)."""
    try:
        from importlib.resources import files

        return (files(_PRICING_PKG) / _PRICING_FILE).read_text(encoding="utf-8")
    except Exception:
        # Fallback for older importlib semantics.
        from importlib.resources import read_text

        return read_text(_PRICING_PKG, _PRICING_FILE)


def _dec(value) -> Decimal:
    # Convert through str to avoid binary-float artifacts.
    return Decimal(str(value))


class _ModelRule:
    __slots__ = (
        "provider",
        "model",
        "match",
        "operation",
        "input",
        "cached_input",
        "output",
        "reasoning",
        "tool",
        "image_tiers",
        "audio_per_second",
        "per_character",
        "batch_discount",
        "finetune_surcharge",
    )

    def __init__(self, raw: dict):
        self.provider: str = raw["provider"]
        self.model: str = raw["model"]
        self.match: str = raw.get("match", "exact")
        self.operation: str = raw.get("operation", "chat")
        self.input = _dec(raw["input_per_token"]) if "input_per_token" in raw else None
        self.cached_input = (
            _dec(raw["cached_input_per_token"])
            if "cached_input_per_token" in raw
            else None
        )
        self.output = (
            _dec(raw["output_per_token"]) if "output_per_token" in raw else None
        )
        self.reasoning = (
            _dec(raw["reasoning_per_token"]) if "reasoning_per_token" in raw else None
        )
        self.tool = _dec(raw["tool_per_token"]) if "tool_per_token" in raw else None
        self.image_tiers: List[dict] = raw.get("image_tiers", []) or []
        self.audio_per_second = (
            _dec(raw["audio_per_second"]) if "audio_per_second" in raw else None
        )
        self.per_character = (
            _dec(raw["per_character"]) if "per_character" in raw else None
        )
        self.batch_discount = (
            _dec(raw["batch_discount"]) if "batch_discount" in raw else None
        )
        self.finetune_surcharge = (
            _dec(raw["finetune_surcharge"]) if "finetune_surcharge" in raw else None
        )

    def matches(self, provider: str, model: str) -> bool:
        if self.provider != provider:
            return False
        if self.match == "prefix":
            return model.startswith(self.model)
        return model == self.model


class PricingEngine:
    """Loads pricing rules and computes per-call cost."""

    _DEFAULT: Optional["PricingEngine"] = None
    _LOCK = threading.Lock()

    def __init__(self, yaml_text: Optional[str] = None):
        import yaml

        text = yaml_text if yaml_text is not None else _load_pricing_text()
        data = yaml.safe_load(text)
        self.version: str = str(data.get("version", ""))
        self.currency: str = data.get("currency", "USD")
        self._rules: List[_ModelRule] = [_ModelRule(m) for m in data.get("models", [])]
        # Cache resolved (provider, model) -> rule lookups.
        self._cache: Dict[Tuple[str, str], Optional[_ModelRule]] = {}

    @classmethod
    def default(cls) -> "PricingEngine":
        if cls._DEFAULT is None:
            with cls._LOCK:
                if cls._DEFAULT is None:
                    cls._DEFAULT = cls()
        return cls._DEFAULT

    @property
    def pricing_version(self) -> str:
        return self.version

    def _find(self, provider: str, model: str) -> Optional[_ModelRule]:
        key = (provider, model)
        if key in self._cache:
            return self._cache[key]
        found: Optional[_ModelRule] = None
        # Exact matches take priority over prefix matches.
        exacts = [r for r in self._rules if r.match == "exact" and r.matches(provider, model)]
        if exacts:
            found = exacts[0]
        else:
            prefixes = [r for r in self._rules if r.match == "prefix" and r.matches(provider, model)]
            if prefixes:
                # Longest prefix wins (most specific).
                found = max(prefixes, key=lambda r: len(r.model))
        self._cache[key] = found
        return found

    def cost(
        self,
        usage: Usage,
        provider: str,
        model: str,
        operation: str = "chat",
        batch: bool = False,
        finetuned: bool = False,
    ) -> Decimal:
        """Compute the cost in USD for one call. Returns Decimal('0') if unknown."""
        rule = self._find(provider, model)
        if rule is None:
            return Decimal("0")

        total = Decimal("0")

        # Token-priced components. Cached tokens are billed at the cached rate
        # and are assumed to be a subset already excluded from input_tokens by
        # the adapters (adapters report non-cached input separately).
        if rule.input is not None:
            total += rule.input * usage.input_tokens
        if usage.cached_input_tokens:
            cached_rate = rule.cached_input if rule.cached_input is not None else rule.input
            if cached_rate is not None:
                total += cached_rate * usage.cached_input_tokens
        if rule.output is not None:
            total += rule.output * usage.output_tokens
        if usage.reasoning_tokens:
            reasoning_rate = rule.reasoning if rule.reasoning is not None else rule.output
            if reasoning_rate is not None:
                total += reasoning_rate * usage.reasoning_tokens
        if usage.tool_tokens and rule.tool is not None:
            total += rule.tool * usage.tool_tokens

        # Audio (per second).
        if usage.audio_seconds and rule.audio_per_second is not None:
            total += rule.audio_per_second * _dec(usage.audio_seconds)

        # Characters (e.g. TTS input text).
        if usage.characters and rule.per_character is not None:
            total += rule.per_character * usage.characters

        # Images (per tile within the first/default tier + flat base per image).
        if usage.image_tiles and rule.image_tiers:
            tier = rule.image_tiers[0]
            per_tile = _dec(tier.get("per_tile", 0))
            base = _dec(tier.get("base", 0))
            total += per_tile * usage.image_tiles
            if usage.image_count:
                total += base * usage.image_count

        # Multipliers.
        if batch and rule.batch_discount is not None:
            total *= rule.batch_discount
        if finetuned and rule.finetune_surcharge is not None:
            total *= rule.finetune_surcharge

        return total
